redfish_client/connector.py

```diff
@@ -37,11 +51,94 @@ def _url(self, path):
 
     def _request(self, method, path, payload=None):
         args = dict(json=payload) if payload else {}
-        return self._client.request(method, self._url(path), **args)
+        resp = self._client.request(method, self._url(path), **args)
 
-    def set_header(self, key, value):
+        if resp.status_code == 401:
+            self.login()
+            resp = self._client.request(method, self._url(path), **args)
+
+        try:
+            json_data = resp.json()
+        except json.JSONDecodeError:
+            json_data = None
+        headers = dict(resp.headers.lower_items())
+
+        return Response(resp.status_code, headers, json_data, resp.content)
+
+    def _set_header(self, key, value):
         self._client.headers[key] = value
 
+    def _unset_header(self, key):
+        if key in self._client.headers:
+            del self._client.headers[key]
+
+    def set_session_auth_data(self, path, session_id=None, token=None):
+        self._basic_logout()
+        self._basic_path = None
+
+        self._session_path = path
+        self._session_id = session_id
+        if token:
+            self._set_header("x-auth-token", token)
+
+    def set_basic_auth_data(self, path):
+        self._session_logout()
+        self._session_path = None
+
+        self._basic_path = path
+
+    @property
+    def _has_session_support(self):
+        return bool(self._session_path)
+
+    def _session_login(self):
+        resp = self._client.post(self._url(self._session_path), json=dict(
+            UserName=self._username, Password=self._password,
+        ))
+        if resp.status_code != 201:
+            raise AuthException("Cannot create session: {}".format(resp.text))
+
+        self._set_header("x-auth-token", resp.headers["x-auth-token"])
+        # We combine with `or` here because the default value of the dict.get
+        # method is eagerly evaluated, which is not what we want.
+        self._session_id = (
+            resp.headers.get("location") or resp.json()["@odata.id"]
+        )
+
+    def _session_logout(self):
+        if self._session_id:
+            self._client.delete(self._url(self._session_id))
+            self._session_id = None
+        self._unset_header("x-auth-token")
+
+    def _basic_login(self):
+        secret = "Basic {}".format(base64.b64encode(
+            "{}:{}".format(self._username, self._password).encode("ascii"),
+        ).decode("ascii"))
+        resp = self._client.get(
+            self._url(self._basic_path), headers=dict(authorization=secret),
+        )
+        if resp.status_code != 200:
+            raise AuthException("Invalid credentials")
+        self._set_header("authorization", secret)
+
+    def _basic_logout(self):
+        self._unset_header("authorization")
+
+    def login(self):
+        assert self._session_path or self._basic_path, "Use set_*_auth_data"
+
+        if self._has_session_support:
+            self._basic_logout()
```

Comment:
I'm not certain if we want this here. Doesn't it feel a bit uncommon to send DELETE request upon login? Not sure what your usecase is, but I'd keep login pure, if user wants to logout first, she can invoke `loguot` anyway.